cql3/expr/expression.cc

```diff
@@ -1411,8 +1415,172 @@ constant evaluate(const expression& e, const query_options& options) {
     }, e);
 }
 
+// Checks whether a bound value of this type has to be reserialized
+static bool needs_to_be_reserialized(const abstract_type& type, const cql_serialization_format& sf) {
+    // If a value contains set or map, then this collection can be sent in the wrong order
+    // or there could be duplicates inside. We need to reserialize it into proper set or map.
+    if (type.contains_set_or_map()) {
+        return true;
+    }
+
+    // If a value contains a collection and it's serialized using the old serialization format,
+    // then we need to reserialize the collection to the current serialization format.
+    if (!sf.using_32_bits_for_collections() && type.contains_collection()) {
+        return true;
+    }
+
+    return false;
+}
+
+// Serializes a map using the internal cql serialization format
+// Takes a range of pair<managed_bytes, managed_bytes>
+template <typename Range>
+requires requires (Range map_range) {
+    {*map_range.begin()} -> std::convertible_to<const std::pair<managed_bytes, managed_bytes>&>;
+}
```

Comment:
You can write:

```c__
template <std::ranges::range Range>
requires std::convertible_to<Range, const std::pair<...>>
```

Reply:
Don't we have such code already?

Reply:
> You can write: `requires std::convertible_to<Range, const std::pair<...>>`

Range is in this case `std::map`, it can't be convertible to `std::pair`

I changed it to:
```c++
template <std::ranges::range Range>
requires requires (Range map_range) {
    {*std::ranges::begin(map_range)} -> std::convertible_to<const std::pair<managed_bytes, managed_bytes>&>;
}
```
I didn't do much concepts, so there might be some easier way that I'm not seeing

> Don't we have such code already?

The existing code does it the same way:
https://github.com/scylladb/scylla/blob/dd9d6c081ee48e670eda70e14a765442db62c0ec/cql3/maps.cc#L88

I looked around a bit but didn't find anything I could use for this

Reply:
Map's value_type is `std::pair<const Key, Value>`. Note the const.

Reply:
And I should have written:

```c++
template <std::ranges::range Range>
requires std::convertible_to<std::ranges::range_value_t<Range>, std::pair<...>>
```

Reply:
That's nice, changed.